stdlib/SparseArrays/src/linalg.jl

```diff
@@ -1,12 +1,44 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-import LinearAlgebra: checksquare
+import LinearAlgebra: checksquare, sym_uplo
 using Random: rand!
 
 # In matrix-vector multiplication, the correct orientation of the vector is assumed.
 const StridedOrTriangularMatrix{T} = Union{StridedMatrix{T}, LowerTriangular{T}, UnitLowerTriangular{T}, UpperTriangular{T}, UnitUpperTriangular{T}}
 const AdjOrTransStridedOrTriangularMatrix{T} = Union{StridedOrTriangularMatrix{T},Adjoint{<:Any,<:StridedOrTriangularMatrix{T}},Transpose{<:Any,<:StridedOrTriangularMatrix{T}}}
 
+for HermSym ∈ [:Hermitian, :Symmetric]
+    @eval begin
+        function (+)(A::SparseMatrixCSC, B::$HermSym{<:Any,<:AbstractSparseMatrix})
+            C  = sparse(B)
+            C .= A .+ C
```

Comment:
Doing things in place seems to be more efficient:
```julia
julia> using SparseArrays

julia> A, B, C = sprand(10000, 10000, 0.05), sprand(10000, 10000, 0.05), spzeros(10000, 10000);

julia> @btime A + B;
  63.968 ms (8 allocations: 152.66 MiB)

julia> @btime C .= A .+ B;
  61.271 ms (2 allocations: 48 bytes)
```

Regarding explicit return statements, the majority of functions in this file don't use them, and personally I dislike the `return` keyword.

Reply:
Ok, I wasn't sure if there really is an "in-place" implementation. The 48 bytes don't look reasonable to me: the vector of values of `C` initially doesn't have any entries, but needs to store all entries of `A .+ B` after the addition. And the difference in runtime (given the large matrix size and the many nonzero entries) seems negligible to me. But I don't know for sure. We'll get some advice later, probably @mbauman or @andreasnoack can help with this issue, currently it doesn't affect functionality, so we can start testing this in any case.

Reply:
It's because BenchmarkTools repeatedly assigns into the _same_ `C`, so you're not really measuring the time-from-empty.  Also should be using `@btime $A + $B` for more accurate answers.  You'll need to be more careful about the problem `setup` in order to get a good grasp here.

Reply:
Right, my bad (I swear I'm usually more careful about benchmarking!):
```julia
julia> @benchmark Cz .= $A .+ $B setup=(Cz = spzeros(10000, 10000))
BenchmarkTools.Trial: 
  memory estimate:  152.54 MiB
  allocs estimate:  2
  --------------
  minimum time:     65.595 ms (0.00% GC)
  median time:      94.922 ms (0.00% GC)
  mean time:        96.968 ms (2.74% GC)
  maximum time:     117.400 ms (17.02% GC)
  --------------
  samples:          50
  evals/sample:     1

julia> @benchmark $A .+ $B
BenchmarkTools.Trial: 
  memory estimate:  152.62 MiB
  allocs estimate:  8
  --------------
  minimum time:     78.753 ms (0.00% GC)
  median time:      91.994 ms (0.00% GC)
  mean time:        98.202 ms (5.81% GC)
  maximum time:     112.802 ms (15.47% GC)
  --------------
  samples:          51
  evals/sample:     1

```

Reply:
Apologies for the poor benchmarks, I think this is more representative:
```julia
julia> @benchmark B .= A .+ B setup=((A = sprandn(1000, 1000, 0.05); B = sprandn(1000, 1000, 0.05))) seconds=10
BenchmarkTools.Trial: 
  memory estimate:  1.91 MiB
  allocs estimate:  8
  --------------
  minimum time:     689.281 μs (0.00% GC)
  median time:      746.601 μs (0.00% GC)
  mean time:        765.545 μs (2.04% GC)
  maximum time:     2.161 ms (60.23% GC)
  --------------
  samples:          3951
  evals/sample:     1

julia> @benchmark A + B setup=((A = sprandn(1000, 1000, 0.05); B = sprandn(1000, 1000, 0.05))) seconds=10
BenchmarkTools.Trial: 
  memory estimate:  1.52 MiB
  allocs estimate:  7
  --------------
  minimum time:     587.142 μs (0.00% GC)
  median time:      596.713 μs (0.00% GC)
  mean time:        609.592 μs (1.45% GC)
  maximum time:     1.551 ms (58.68% GC)
  --------------
  samples:          4266
  evals/sample:     1
```
which suggests that @dkarrasch's suggestion is the way to go.